department-of-veterans-affairs/caseflow
# Lianna/APPEALS-12481

Resolves APPEALS-12481

### Description
NonCompTab.jsx
1. Update to have access to parameters that it doesn't currently have. Current usage is of tasks which we will no longer be using. Instead we need to be passing the TaskPages api endpoint to the TaskTable (potentially via the erb file that gets passed in as a prop)
2. Have to create and pass down the pagination options. Those will either be retrieved from the GET params in the url or created in NonCompTabs and passed down to the TaskTable. 

TaskTable.jsx
1. Make sure that the parameters coming from NonCompTabs are passed down properly through TaskTable. This is because the TaskTable renders the QueueTable which is ultimately responsible for much of the UI behavior. Properties needed for the api call to run successfully:
`useTaskPagesApi`
`tabPaginationOptions= {this.props.paginationOptions}`
`taskPagesApiEndpoint= {this.props.taskPagesApiEndpoint}`
2. Default sort is currently not showing through the UI even though it is currently working

QueueTable.jsx
1. We need to set that property so it can be sent to the server in the `requestQueryString` method. `validatedPaginationOptions` is responsible for setting the state, so we'd need to pull the parameter out from there to set the pagination options

utils.js
1. This file is used by QueueTable when it retrieves data from the server. Specifically it uses `tasksWithAppealFromRawTasks` which utilizes other utils.js functions to create a JS object from the tasks received from a controller. For the DR queue, this data will be retrieved from the `DecisionReviewTaskSerializer`. Currently the serialized object makes assumptions on the task data it receives and will fail because they don't use the `?.` safe operator

TaskTableColumns.jsx (2 - one in NonComp, one in Queue)
1. Attribute references in these files may require changes due to updates both to the `DecisionReviewTaskSerializer` and `utils.js`. Any updates to these columns would need to be accounted for in places where they are currently in usage, as pagination updates would likely introduce breaking changes

### Acceptance Criteria

- [ ] Update [NonCompTabs](https://github.com/department-of-veterans-affairs/caseflow/blob/dr-queue-pagination-research/client/app/nonComp/components/NonCompTabs) to allow for new params to be passed
- [ ] Update [TaskTable](https://github.com/department-of-veterans-affairs/caseflow/blob/master/client/app/queue/components/TaskTable.jsx) to allow for params to be passed from NonCompTabs
- [ ] Update attribute references in [NonComp TaskTableColumns](https://github.com/department-of-veterans-affairs/caseflow/blob/master/client/app/nonComp/components/TaskTableColumns.jsx)
- [ ] Update attribute references in [Queue TaskTableColumns](https://github.com/department-of-veterans-affairs/caseflow/blob/master/client/app/queue/components/TaskTableColumns.jsx)

### Testing Plan
1. While the changes are on this branch, in order to run the react tests and see the front end changes you'll need to combine this story with it's backend counter part APPEALS-12483. Matt has already created a combined branch to aid in testing. Begin by fetching all the latest changes and switching from branch `Lianna/APPEALS-12481` to branch `12481-and12483-combined`.
2. Once on the `12481-and12483-combined` branch move into the client folder for running the react testing suite `cd client`
3. Once in the client folder run the tests with ` yarn jest test/app/nonComp/NonCompTabs.test.js`
4. The testing suite should show two tests for NonCompTabs and both tests should pass.

<img width="1422" alt="Running NonCompTabs Test" src="https://user-images.githubusercontent.com/17394371/209350532-b2d26ee9-3872-476a-b71b-0ca6ecfb1bdf.png">

<img width="879" alt="NonCompTabs Tests Passing" src="https://user-images.githubusercontent.com/17394371/209350605-8300f4ea-4a44-491f-860b-19dfdc779e53.png">

5. Once confirmed that tests are passing confirm the implementation works with the backend

6. While running frontend and backend of application go to the browser and type localhost:3000 and login with user  `ADMININTAKE_LOCAL (VACO)`

7. Navigate to  `http://localhost:3000/decision_reviews/vha` and confirm you are on the _VHA Decisions Review Page_
<img width="1332" alt="VHA Decision Review Page" src="https://user-images.githubusercontent.com/17394371/209361824-2766e2ce-65b6-4901-b70f-de4a874b3f45.png">

8. Make sure there are enough listed tasks to go to at least page 3, if not, go to terminal > then rails console and add a few more tasks by running the following:
- 1. `cd ..` (return to main caseflow folder from client folder used for testing)
- 2. `make c`
- 3. `30.times do
        new_decision_task = FactoryBot.create(:higher_level_review_vha_task)
       end`

- 4. Return to browser and refresh the page

- 5. Once more page numbers appear you can move to step nine

9. Confirm that refreshing the page keeps you on the same page

10. Go to Page 3 

11. Confirm that refreshing the page keeps you on page 3
<img width="1680" alt="Page 3 persists" src="https://user-images.githubusercontent.com/17394371/209363224-e1861144-7fe6-422b-9fef-173bfcfa21f0.png">

12. Go to Completed tasks tab and refresh page, confirm you are still in the Completed tasks tab.

<img width="1680" alt="Completed Tasks Tab" src="https://user-images.githubusercontent.com/17394371/209364400-d77903a0-325e-400b-9287-d43be056ce53.png">
